tpu/hptuning/resnet-hptuning/resnet/resnet_main.py

```diff
@@ -0,0 +1,512 @@
+          lr: `Tensor` with shape `[batch]` for the learning_rate.
+          ce: `Tensor` with shape `[batch]` for the current_epoch.
+
+        Returns:
+          List of summary ops to run on the CPU host.
+        """
+        gs = gs[0]
+        with summary.create_file_writer(FLAGS.model_dir).as_default():
+          with summary.always_record_summaries():
+            summary.scalar('loss', loss[0], step=gs)
+            summary.scalar('learning_rate', lr[0], step=gs)
+            summary.scalar('current_epoch', ce[0], step=gs)
+
+            return summary.all_summary_ops()
+
+      # To log the loss, current learning rate, and epoch for Tensorboard, the
+      # summary op needs to be run on the host CPU via host_call. host_call
+      # expects [batch_size, ...] Tensors, thus reshape to introduce a batch
+      # dimension. These Tensors are implicitly concatenated to
+      # [params['batch_size']].
+      gs_t = tf.reshape(global_step, [1])
+      loss_t = tf.reshape(loss, [1])
+      lr_t = tf.reshape(learning_rate, [1])
+      ce_t = tf.reshape(current_epoch, [1])
+
+      host_call = (host_call_fn, [gs_t, loss_t, lr_t, ce_t])
+
+  else:
+    train_op = None
+
+  eval_metrics = None
+  if mode == tf.estimator.ModeKeys.EVAL:
+    def metric_fn(labels, logits):
+      """Evaluation metric function. Evaluates accuracy.
+
+      This function is executed on the CPU and should not directly reference
+      any Tensors in the rest of the `model_fn`. To pass Tensors from the model
+      to the `metric_fn`, provide as part of the `eval_metrics`. See
+      https://www.tensorflow.org/api_docs/python/tf/contrib/tpu/TPUEstimatorSpec
+      for more information.
+
+      Arguments should match the list of `Tensor` objects passed as the second
+      element in the tuple passed to `eval_metrics`.
+
+      Args:
+        labels: `Tensor` with shape `[batch]`.
+        logits: `Tensor` with shape `[batch, num_classes]`.
+
+      Returns:
+        A dict of the metrics to return from evaluation.
+      """
+      predictions = tf.argmax(logits, axis=1)
+      top_1_accuracy = tf.metrics.accuracy(labels, predictions)
+      in_top_5 = tf.cast(tf.nn.in_top_k(logits, labels, 5), tf.float32)
+      top_5_accuracy = tf.metrics.mean(in_top_5)
+
+      return {
+          'top_1_accuracy': top_1_accuracy,
+          'top_5_accuracy': top_5_accuracy,
+      }
+
+    eval_metrics = (metric_fn, [labels, logits])
+
+  return tpu_estimator.TPUEstimatorSpec(
+      mode=mode,
+      loss=loss,
+      train_op=train_op,
+      host_call=host_call,
+      eval_metrics=eval_metrics)
+
+
+def main(unused_argv):
+  trial_id = os.environ.get('CLOUD_ML_TRIAL_ID', 0)
+  if trial_id:
+    if FLAGS.export_dir:
+      FLAGS.export_dir = os.path.join(
+        FLAGS.export_dir, 'trial_{}'.format(trial_id))
+    FLAGS.model_dir = os.path.join(
+      FLAGS.model_dir, 'trial_{}'.format(trial_id))
+
+  tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(
```

Comment:
This can benefit from explaining why TPUCLusterResolver is needed (as in difference in workflow compared to CPU/GPU). Or that will be part of blog?